### PR TITLE
Regularization

### PR DESCRIPTION
**Pull request by mnwright**
Continuing #457 and #462.

### CODE REVIEW

File: src/TreeRegression.cpp [hunk 1]
```diff
@@ -862,25 +915,54 @@ void TreeRegression::findBestSplitValueBeta(size_t nodeID, size_t varID, double
...
+    }
+
+    if ((*all_split_varIDs)[best_varID] != 1){
+      if (use_depth == 1){  
+        best_decrease = best_decrease * std::pow(coef_reg[best_varID], next_depth);
+      } else {
+        best_decrease = best_decrease  * coef_reg[best_varID];
+      }
+    }
+
```
Comment (mnwright) [inline src/TreeRegression.cpp]: Why do we need this?

**Comment by mnwright** [comment]
TODOS (for myself): 

- Fix regularization if corrected variable importance used (use unpermuted varID)
- Add survival forests
- Only reserve memory etc. when regularization used
- Use member variables, remove parameters from function calls

File: src/TreeRegression.cpp (hunk 1 above)
Comment (brunaw) [reply src/TreeRegression.cpp]: Hey Marvin. That's for keeping the variable importance values on the same scale of a non-regularized model. So if we run a regularized and a non-regularized random forest, their importance values will be similar (and comparable), with the difference that the regularized model will be using fewer variables. I think it's a nice feature to keep, though looking now it seems like could be better written

Comment (mnwright) [reply src/TreeRegression.cpp]: Thanks. Why do you penalize two times? Wouldn't the first if-clause be enough? Then the decrease of Gini impurity would be 

(s1/n1 + s2/n2) * reg - s/n * reg

where _s_ is the squared sum in the current node, _s1_ in the left child node and _s2_ in the right child node, same for the sample sizes _n_, _reg_ is the regularization factor. 

I think currently you are multiplying by the regularization factor again: 

((s1/n1 + s2/n2) * reg - s/n * reg) * reg

Is that on purpose?

Comment (brunaw) [reply src/TreeRegression.cpp]: That's true, the factor is being used two times. I don't recall any reason why that should be used, sorry about that! so the first if-clause should be enough. Do you mind changing this part?

Comment (mnwright) [reply src/TreeRegression.cpp]: No problem, I will change it.

Comment (brunaw) [reply src/TreeRegression.cpp]: Thanks Marvin!

**Comment by mnwright** [comment]
I decided to change the interface. It's now: 

````R
library(ranger)
ranger(mpg ~ ., data = mtcars, num.threads = 1,
      regularization.factor = 0.1, 
      regularization.usedepth = TRUE)
````

That's a little bit long, but consistent with other parameters (we don't have any other complex parameters with lists etc..)

@brunaw: Do you think we should use the depth by default?

**Comment by brunaw** [comment]
Thanks Marvin! That makes sense to me. 

No, I think the use of the depth should be consciously set by the user. In my experiments, using the depth leads to good models, but sometimes it can lead to "too much regularized" models (using almost no variables) which compromises the prediction power, and in this case, it might be better not to use the depth at all. I would like the user to be aware of when this happens :)

**Comment by mnwright** [comment]
I see. However, we need to set a default and that would mean the default is `FALSE`. That also makes sense somehow because we can cite [Deng & Runger](https://arxiv.org/abs/1201.1587) for this.

**Comment by brunaw** [comment]
Perfect, thanks Marvin!

**Pull request merged**